File: shezle_stock/shezle_stock/report/item_stock_balance_report/item_stock_balance_report.py
import frappe

def execute(filters=None):
    """
    Main function to execute the report with the given filters.
    """
    if not filters:
        filters = {}
    
    item_name = filters.get('item')
    periodicity = filters.get('periodicity')

    columns = get_columns()
    data = get_data_updated(item_name, periodicity)
    
    return columns, data

def get_columns():
    """
    Define the columns for the report.
    """
    return [
        {
            'fieldname': 'item_code_number',
            'label': 'Item Code Number',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'item_name',
            'label': 'Item Name',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'item_code',
            'label': 'Item Code Name (One of multiple)',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'barcode',
            'label': 'Barcode No',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'stock_uom',
            'label': 'Stock Unit',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'item_group',
            'label': 'Item Group',
            'fieldtype': 'Link',
            'options': "Item Group"
        },
        {
            'fieldname': 'warehouse_name',
            'label': 'Warehouse Name',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'location_name',
            'label': 'Location Name',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'qty',
            'label': 'Qty',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'effective_unit_rate',
            'label': 'Effective Unit Rate',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'value',
            'label': 'Value',
            'fieldtype': 'Data',
        },
    ]

def get_data_updated(item_name, periodicity):
    """
    Fetch and filter data based on item name and periodicity.
    """
    conditions = ""
    start_date = None

    # Determine start date based on periodicity
    if periodicity:
        today = frappe.utils.nowdate()
        if periodicity == 'Monthly':
            start_date = frappe.utils.add_months(today, -1)
        elif periodicity == 'Quarterly':
            start_date = frappe.utils.add_months(today, -3)
        elif periodicity == 'Half-Yearly':
            start_date = frappe.utils.add_months(today, -6)
        elif periodicity == 'Yearly':
            start_date = frappe.utils.add_years(today, -1)

    # Add start date condition if available
    if start_date:
        conditions += f" AND it.modified >= '{start_date}'"

    # Add item name condition if available
    if item_name:
        conditions += f" AND pii.item_code = '{item_name}'"

    # SQL query to fetch data
    query = f"""
        SELECT 
    pii.item_code, 
    it.barcode AS barcode,
    it.item_group, 
    it.modified,
    it.variant_of AS item_code_number,
    pii.uom AS stock_uom, 
    SUM(pii.qty) AS qty,
    ROUND(SUM(pii.qty * pii.rate) / SUM(pii.qty), 2) AS effective_unit_rate, 
    SUM(pii.qty) * ROUND(SUM(pii.qty * pii.rate) / SUM(pii.qty), 2) AS value
FROM 
    `tabPurchase Invoice Item` pii
INNER JOIN 
    `tabPurchase Invoice` pi ON pi.name = pii.parent AND pi.update_stock = 1
LEFT JOIN 
    (SELECT 
        item.item_code, 
        item_barcode.barcode, 
        item.item_group, 
        item.variant_of,
        item.modified
    FROM 
        `tabItem` item 
    LEFT JOIN 
        `tabItem Barcode` item_barcode ON item.name = item_barcode.parent) it
ON 
    pii.item_code = it.item_code
WHERE 
    1=1
    {conditions} 
GROUP BY 
    it.item_code;

    """

    # Execute the query
    data = frappe.db.sql(query, as_dict=True)
    
    # Add item name to each entry in the data
    for entry in data:
        entry['item_name'] = frappe.db.get_value("Item", entry['item_code_number'], 'item_name')
    
    return data

File: shezle_stock/shezle_stock/report/item_stock_balance_report/item_stock_balance_report.js
// Copyright (c) 2024, yes and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["item Stock Balance report"] = {
	filters: [
        {
            fieldname: 'periodicity',
            label: __('Periodicity'),
            fieldtype: 'Select',
            options: ['Monthly', 'Quarterly', 'Half-Yearly', 'Yearly', 'Custom'],
            default: 'Yearly',
            change: () => {
                const today = frappe.datetime.get_today();
                const periodicity = frappe.query_report.get_filter_value('periodicity');
                console.log(periodicity)
                
                if (periodicity === 'Monthly') {
                    frappe.query_report.set_filter_value('to_date',today)
                    let month1 = frappe.datetime.add_months(today, -1);
                    frappe.query_report.set_filter_value('from_date',month1)

                } else if (periodicity === 'Quarterly') {
                    frappe.query_report.set_filter_value('to_date',today)
                    let month1 = frappe.datetime.add_months(today, -3);
                    frappe.query_report.set_filter_value('from_date',month1)

                } else if (periodicity === 'Half-Yearly') {
                    frappe.query_report.set_filter_value('to_date',today)
                    let month1 = frappe.datetime.add_months(today, -6);
                    frappe.query_report.set_filter_value('from_date',month1)
                } else if (periodicity === 'Yearly') {
                    frappe.query_report.set_filter_value('to_date',today)
                    let month1 = frappe.datetime.add_months(today, -12);
                    frappe.query_report.set_filter_value('from_date',month1)
                } else {
                    // Handle 'Custom' periodicity or default to today's date
                    return today;
                }
            }
        },
        {
            fieldname: 'from_date',
            label: __('From Date'),
            fieldtype: 'Date',
            // change: () => {
                
            //     frappe.query_report.set_filter_value('periodicity','Custom');
                
            //     console.log(periodicity)
                
                
            // }
        
            

            
        },
        {
            fieldname: 'to_date',
            label: __('To Date'),
            fieldtype: 'Date',
            default: frappe.datetime.get_today(),  // Default to today's date
        }
		
		
	],
        
    
};
